cpp/src/arrow/ipc/feather.h

```diff
@@ -116,55 +93,42 @@ class ARROW_EXPORT TableReader {
   /// \return Status
   ///
   /// This function is zero-copy if the file source supports zero-copy reads
-  Status Read(const std::vector<std::string>& names, std::shared_ptr<Table>* out);
-
- private:
-  class ARROW_NO_EXPORT TableReaderImpl;
-  std::unique_ptr<TableReaderImpl> impl_;
+  virtual Status Read(const std::vector<std::string>& names,
+                      std::shared_ptr<Table>* out) = 0;
 };
 
-/// \class TableWriter
-/// \brief Interface for writing Feather files
-class ARROW_EXPORT TableWriter {
- public:
-  ~TableWriter();
-
-  /// \brief Create a new TableWriter that writes to an OutputStream
-  /// \param[in] stream an output stream
-  /// \param[out] out the returned table writer
-  /// \return Status
-  static Status Open(const std::shared_ptr<io::OutputStream>& stream,
-                     std::unique_ptr<TableWriter>* out);
+struct ARROW_EXPORT WriteProperties {
+  static WriteProperties Defaults() { return WriteProperties(); }
 
-  /// \brief Set the description field in the file metadata
-  void SetDescription(const std::string& desc);
+  static WriteProperties DefaultsV1() {
+    WriteProperties props;
+    props.version = kFeatherV1Version;
+    return props;
+  }
 
-  /// \brief Set the number of rows in the file
-  void SetNumRows(int64_t num_rows);
-
-  /// \brief Append a column to the file
+  /// Feather file version number
   ///
-  /// \param[in] name the column name
-  /// \param[in] values the column values as a contiguous arrow::Array
-  /// \return Status
-  Status Append(const std::string& name, const Array& values);
+  /// version 2: "Feather V1" Apache Arrow <= 0.16.0
+  /// version 3: "Feather V2" Apache Arrow > 0.16.0
+  int version = kFeatherV2Version;
 
-  /// \brief Write a table to the file
-  ///
-  /// \param[in] table the table to be written
-  /// \return Status
-  Status Write(const Table& table);
+  // Parameters for Feather V2 only
 
-  /// \brief Finalize the file by writing the file metadata and footer
-  /// \return Status
-  Status Finalize();
+  /// Number of rows per intra-file chunk. Use smaller chunksize when you need
+  /// faster random row access
+  int64_t chunksize = 1LL << 16;
 
- private:
-  TableWriter();
-  class ARROW_NO_EXPORT TableWriterImpl;
-  std::unique_ptr<TableWriterImpl> impl_;
+  /// Compression type to use
+  Compression::type compression = Compression::ZSTD;
```

Comment:
Is this correct: zstd compressed by default? Since `ARROW_IPC` does not require `ARROW_WITH_ZSTD`, the default may very well fail (as it turns out from what I'm seeing locally).

Note that the python bindings (docstring at least) say uncompressed is default.

Reply:
Ah I'll fix that. If anything LZ4 should probably be the default since based on benchmarks it has little effect on read/write times.

I'll change the implementation to set LZ4 as the default if it's available otherwise uncompressed is the default.

Reply:
Done